mlos_core/mlos_core/optimizers/optimizer.py

```diff
@@ -73,201 +75,182 @@ def space_adapter(self) -> Optional[BaseSpaceAdapter]:
+        ), "Only one of observations or observation can be provided."
 
-        context : pd.DataFrame
-            Not Yet Implemented.
+        if observation is not None:
+            self.register_single(observation=observation)
+        if observations is not None:
+            for obs in observations.to_list():
+                self.register_single(observation=obs)
 
-        metadata : Optional[pd.DataFrame]
-            Not Yet Implemented.
+    def register_single(self, *, observation: Observation) -> None:
+        """
+        Wrapper method, which employs the space adapter (if any), before registering the
+        configs and scores.
+
+        Parameters
+        ----------
+        observation: Observation
+            The observation to register.
         """
         # Do some input validation.
-        assert metadata is None or isinstance(metadata, pd.DataFrame)
-        assert set(scores.columns) == set(
+        assert observation.metadata is None or isinstance(observation.metadata, pd.Series)
+        assert set(observation.score.index) == set(
             self._optimization_targets
         ), "Mismatched optimization targets."
         assert self._has_context is None or self._has_context ^ (
-            context is None
+            observation.context is None
         ), "Context must always be added or never be added."
-        assert len(configs) == len(scores), "Mismatched number of configs and scores."
-        if context is not None:
-            assert len(configs) == len(context), "Mismatched number of configs and context."
-        assert configs.shape[1] == len(
+        assert len(observation.config) == len(
             self.parameter_space.values()
         ), "Mismatched configuration shape."
-        self._observations.append((configs, scores, context))
-        self._has_context = context is not None
 
+        self._has_context = observation.context is not None
+        self._observations.append(observation)
+
+        register_observation = deepcopy(observation)  # Needed to support named tuples
         if self._space_adapter:
-            configs = self._space_adapter.inverse_transform(configs)
-            assert configs.shape[1] == len(
+            register_observation.config = self._space_adapter.inverse_transform(
+                register_observation.config
+            )
+            assert len(register_observation.config) == len(
                 self.optimizer_parameter_space.values()
             ), "Mismatched configuration shape after inverse transform."
-        return self._register(configs=configs, scores=scores, context=context)
+
+        return self._register(observation=register_observation)
 
     @abstractmethod
     def _register(
         self,
         *,
-        configs: pd.DataFrame,
-        scores: pd.DataFrame,
-        context: Optional[pd.DataFrame] = None,
-        metadata: Optional[pd.DataFrame] = None,
+        observation: Observation,
```

Comment:
See comment above about accepting multiple here and leaving the implementation of whether to loop or not up to the backend.